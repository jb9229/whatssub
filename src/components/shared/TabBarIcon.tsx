import React from 'react';
import { StyleSheet } from 'react-native';
import * as Icon from '@expo/vector-icons';

// Static Variable
const DEFAULT_ICON_SIZE = 32;
const COLOR_TABICON_DEFAULT = '#c6cbcf';
const COLOR_TABICON_SELECTED = '#75a3ff';

const styles = StyleSheet.create({
  icon: {
    marginBottom: -3,
  }
});

interface Props {
  name: string;
  focused: boolean;
  type?: string;
  size?: number;
  testID?: string;
}

function Shared(props: Props) {
  const iconSize = props.size || DEFAULT_ICON_SIZE;

  if(props.type && props.type === 'MaterialIcons') {
    return (
      <Icon.MaterialIcons
        testID={props.testID}
        name={props.name}
        size={iconSize}
        style={styles.icon}
        color={props.focused ? COLOR_TABICON_SELECTED : COLOR_TABICON_DEFAULT}
      />
    );
  }

  if(props.type && props.type === 'AntDesign') {
    return (
      <Icon.AntDesign
        testID={props.testID}
        name={props.name}
        size={iconSize}
        style={styles.icon}
        color={props.focused ? COLOR_TABICON_SELECTED : COLOR_TABICON_DEFAULT}
      />
    );
  }

  return (
    <Icon.Ionicons
      testID={props.testID}
      name={props.name}
      size={iconSize}
      style={styles.icon}
      color={props.focused ? COLOR_TABICON_SELECTED : COLOR_TABICON_DEFAULT}
    />
  );
}

export default Shared;
