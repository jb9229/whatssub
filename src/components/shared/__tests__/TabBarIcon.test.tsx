import 'react-native';
import * as React from 'react';
import { RenderResult, render } from '@testing-library/react-native';
import TabBarIcon from '../TabBarIcon';
import renderer from 'react-test-renderer';
// Note: test renderer must be required after react-native.

let props: any;
let component: React.ReactElement;
let testingLib: RenderResult;

const createTestProps = (obj: object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...obj,
});

describe('[TabBarIcon] UI Render Test', () => {
  beforeEach(() => {
    props = createTestProps({ });
    component = (
      <TabBarIcon {...props} testID="tabBarIcon" name="home" focused={false} type="MaterialIcons" />
    );
  });

  it('renders without crashing and snapshot matches', () => {
    const rendered: renderer.ReactTestRendererJSON | null = renderer.create(component).toJSON();
    expect(rendered).toMatchSnapshot(); // Guarantee UI Changed
    expect(rendered).toBeTruthy(); // Not (false, 0, '', null, undefined, and NaN)
  });
});

describe('[TabBarIcon] Interactions Test', () => {
  beforeEach(() => {
    testingLib = render(component);
  });

  it('should be changed the color when it is focused', () => {
    // testingLib.debug();
    /**
     * Unfortunately I have not found a test method.
     *
     * https://testing-library.com/docs/native-testing-library/example-intro
     * https://reactjs.org/docs/test-renderer.html#testrenderer
     * https://jestjs.io/docs/en/expect
     * https://github.com/testing-library/react-testing-library/issues/43
     *   ??expect(getByTestId('tabBarIcon')).toHaveAttribute('color', '#c6cbcf')
     * */
  });
});
