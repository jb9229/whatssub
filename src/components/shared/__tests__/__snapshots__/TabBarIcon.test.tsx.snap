// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[TabBarIcon] UI Render Test renders without crashing and snapshot matches 1`] = `<Text />`;

exports[`[TabBarIcon] render test renders without crashing and snapshot matches 1`] = `<Text />`;
